sshd-core/src/main/java/org/apache/sshd/client/keyverifier/KnownHostsServerKeyVerifier.java

```diff
@@ -267,36 +268,63 @@ protected PublicKeyEntryResolver getFallbackPublicKeyEntryResolver() {
     protected boolean acceptKnownHostEntries(
             ClientSession clientSession, SocketAddress remoteAddress, PublicKey serverKey,
             Collection<HostEntryPair> knownHosts) {
-        // TODO allow for several candidates and check if ANY of them matches the key and has 'revoked' marker
-        HostEntryPair match = findKnownHostEntry(clientSession, remoteAddress, knownHosts);
-        if (match == null) {
+
+        List<HostEntryPair> hostMatches = findKnownHostEntries(clientSession, remoteAddress, knownHosts);
+        if (hostMatches.isEmpty()) {
             return acceptUnknownHostKey(clientSession, remoteAddress, serverKey);
         }
 
-        KnownHostEntry entry = match.getHostEntry();
-        PublicKey expected = match.getServerKey();
-        if (KeyUtils.compareKeys(expected, serverKey)) {
-            return acceptKnownHostEntry(clientSession, remoteAddress, serverKey, entry);
+        String serverKeyType = KeyUtils.getKeyType(serverKey);
+
+        List<HostEntryPair> keyMatches = hostMatches.stream()
+                .filter(entry -> serverKeyType.equals(entry.getHostEntry().getKeyEntry().getKeyType()))
+                .filter(k -> KeyUtils.compareKeys(k.getServerKey(), serverKey))
+                .collect(Collectors.toList());
+
+        if (keyMatches.stream()
+                .anyMatch(k -> !acceptKnownHostEntry(clientSession, remoteAddress, serverKey, k.getHostEntry()))) {
+            return false;
         }
 
-        try {
-            if (!acceptModifiedServerKey(clientSession, remoteAddress, entry, expected, serverKey)) {
-                return false;
+        if (!keyMatches.isEmpty()) {
+            return true;
+        }
+
+        List<HostEntryPair> acceptedHostMatches = hostMatches.stream()
+                .filter(k -> acceptKnownHostEntry(clientSession, remoteAddress, serverKey, k.getHostEntry()))
+                .collect(Collectors.toList());
```

Comment:
A `findAny` would be sufficient instead of collecting the list. If none is found, call `acceptUnknownHostKey`, otherwise call `acceptModifiedServerKey()` _once_ with the found entry.

The normal operation with `acceptModifiedServerKey()` is to ask the user. And if the user says "no", it'll return false. There is no need for a loop here going over all the other keys that we know and that the server didn't present. This will only result in the user having to say "no" multiple times.

Also: we know here that `hostMatches` contains no matching key here. All this does is filter out revoked keys. Please name this variable `nonRevokedMatch`.

It might be even clearer to use `.filter(k -> !"revoked".equals(k.getMarker()))` directly here.